kernel/aster-nix/src/util/mod.rs

```diff
@@ -5,37 +5,53 @@ use aster_frame::vm::VmIo;
 use crate::prelude::*;
 pub mod net;
 
-/// copy bytes from user space of current process. The bytes len is the len of dest.
+/// Read bytes into the `dest` buffer from the user space of the current process.
+/// If successful, the `dest` buffer is filled with exact `dest.len` bytes.
 pub fn read_bytes_from_user(src: Vaddr, dest: &mut [u8]) -> Result<()> {
     let current = current!();
     let root_vmar = current.root_vmar();
     Ok(root_vmar.read_bytes(src, dest)?)
 }
 
-/// copy val (Plain of Data type) from user space of current process.
+/// Read a value of `Pod` type from the user space of the current process.
 pub fn read_val_from_user<T: Pod>(src: Vaddr) -> Result<T> {
     let current = current!();
     let root_vmar = current.root_vmar();
     Ok(root_vmar.read_val(src)?)
 }
 
-/// write bytes from user space of current process. The bytes len is the len of src.
+/// Write bytes from the `src` buffer to the user space of the current process. If successful,
+/// the write length will be equal to `src.len`.
 pub fn write_bytes_to_user(dest: Vaddr, src: &[u8]) -> Result<()> {
     let current = current!();
     let root_vmar = current.root_vmar();
     Ok(root_vmar.write_bytes(dest, src)?)
 }
 
-/// write val (Plain of Data type) to user space of current process.
+/// Write `val` to the user space of the current process.
 pub fn write_val_to_user<T: Pod>(dest: Vaddr, val: &T) -> Result<()> {
     let current = current!();
     let root_vmar = current.root_vmar();
     Ok(root_vmar.write_val(dest, val)?)
 }
 
-/// read a cstring from user, the length of cstring should not exceed max_len(include null byte)
+/// Read a C string from the user space of the current process. The length of the string should not exceed max_len (including the null byte).
+/// Additionally, the entire string should be within a single memory mapping and should not cross the mapping boundary.
+///
+/// FIXME: Should we allow a string to cross the mapping boundary?
 pub fn read_cstring_from_user(addr: Vaddr, max_len: usize) -> Result<CString> {
-    let mut buffer = vec![0u8; max_len];
-    read_bytes_from_user(addr, &mut buffer)?;
+    let vm_mapping = {
+        let current = current!();
+        let root_vmar = current.root_vmar();
+        root_vmar.get_vm_mapping(addr)?
+    };
+
+    let max_read_len = {
+        let max_len_in_mapping = vm_mapping.map_to_addr() + vm_mapping.map_size() - addr;
+        max_len.min(max_len_in_mapping)
+    };
+
+    let mut buffer = vec![0u8; max_read_len];
+    vm_mapping.read_bytes(addr - vm_mapping.map_to_addr(), &mut buffer)?;
```

Comment:
Seems not a good idea: strings tend to be much shorter than `max_len` or the length of mapping. Here it reads at least `min(max_len, max_len_in_mapping)`, which may read a lot of useless bytes. We should devise a method to early stop on null terminator while reading.

Also, long `Vec`s are horrible... but such usages scatters through our codebase now. We may design some copy-less patterns in the future.

Reply:
Thanks for your comment!

> strings tend to be much shorter than max_len or the length of mapping. Here it reads at least min(max_len, max_len_in_mapping), which may read a lot of useless bytes. We should devise a method to early stop on the null terminator while reading.

The new implementation solves this problem. It won't copy so many bytes as before. Referring to [the Linux implementation](https://elixir.bootlin.com/linux/v6.0.9/source/lib/strncpy_from_user.c#L28), the new implementation will at first copy a usize(8 bytes) at a time, and then copy a byte at a time, until it finds the final `\0` byte.

Reply:
> Also, long, infallible Vecs are horrible... but such usages scatters through our codebase now. We may design some copy-less patterns in the future.

I quite agree. 

1. For the infallible heap allocation, @cchanging has developed a reservation-based method, but it doesn't seem to be mature enough for use in Asterinas at present.
2. A copy-less design is exactly what we need! We often perform excessive data copying, which undoubtedly impacts performance in both the network and fs subsystems. We have considered implementing `VmReader` and `VmWriter` for `Vmar`, similar to what has been done for `VmFrame` and `VmSegment`. This approach would allow us to delay actual data copying and minimize the introduction of temporary buffers. However, the situation with Vmar is different from that of VmFrame or VmSegment. If you have any insights or ideas regarding this issue, we can discuss it later.